repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/lineage/LineagePreProcessor.java

```diff
@@ -0,0 +1,424 @@
+            } else {
+                return Collections.emptyList();
+            }
+        } catch (Exception e) {
+            return Collections.emptyList();
+        }
+    }
+
+    private AtlasEntity createConnectionProcessEntity(Map<String, Object> connectionProcessInfo) throws AtlasBaseException {
+        AtlasEntity processEntity = new AtlasEntity(CONNECTION_PROCESS_ENTITY_TYPE);
+        processEntity.setAttribute(NAME, connectionProcessInfo.get("connectionProcessName"));
+        processEntity.setAttribute(QUALIFIED_NAME, connectionProcessInfo.get("connectionProcessQualifiedName"));
+        processEntity.setAttribute(HAS_LINEAGE, true);
+
+        AtlasObjectId inputConnection = new AtlasObjectId(CONNECTION_ENTITY_TYPE, mapOf(QUALIFIED_NAME, connectionProcessInfo.get("input")));
+        AtlasObjectId outputConnection = new AtlasObjectId(CONNECTION_ENTITY_TYPE, mapOf(QUALIFIED_NAME, connectionProcessInfo.get("output")));
+
+        Map<String, Object> relationshipAttributes = new HashMap<>();
+        relationshipAttributes.put("inputs", Collections.singletonList(inputConnection));
+        relationshipAttributes.put("outputs", Collections.singletonList(outputConnection));
+        processEntity.setRelationshipAttributes(relationshipAttributes);
+
+        try {
+            RequestContext.get().setSkipAuthorizationCheck(true);
+            AtlasEntity.AtlasEntitiesWithExtInfo processExtInfo = new AtlasEntity.AtlasEntitiesWithExtInfo();
+            processExtInfo.addEntity(processEntity);
+            EntityStream entityStream = new AtlasEntityStream(processExtInfo);
+            entityStore.createOrUpdate(entityStream, false);
+
+            checkAndUpdateConnectionLineage(inputConnection, true);
+            checkAndUpdateConnectionLineage(outputConnection, true);
+        } finally {
+            RequestContext.get().setSkipAuthorizationCheck(false);
+        }
+
+        return processEntity;
+    }
+
+    @Override
+    public void processDelete(AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processDeleteLineageProcess");
+
+        try {
+            Set<String> connectionProcessQNs = getConnectionProcessQNs(vertex);
+            if (connectionProcessQNs.isEmpty()) {
+                return;
+            }
+
+            for (String connectionProcessQn : connectionProcessQNs) {
+                if (!checkIfChildProcessExistForConnectionProcess(connectionProcessQn)) {
+                    deleteConnectionProcess(connectionProcessQn);
+                }
+            }
+        }
+        catch (Exception exp) {
+            LOG.error("Error in LineagePreProcessor.processDelete", exp);
+        }
+        finally {
+            RequestContext.get().endMetricRecord(metricRecorder);
+        }
+    }
+
+    private boolean checkIfChildProcessExistForConnectionProcess(String connectionProcessQn) throws AtlasBaseException {
```

Comment:
This method is always called in loop, this results in running ES query `n` times. Please change logic to call it only once for `n` connectionProcessQns.
You simply provide list of all expected connectionProcessQns in clause mapOf("term", mapOf(PARENT_CONNECTION_PROCESS_QUALIFIED_NAME, connectionProcessQn))
E.g. mapOf("term", mapOf(PARENT_CONNECTION_PROCESS_QUALIFIED_NAME, connectionProcessQns))

ES considers it as `OR` & will return all matching Processes in single call, later you can iterate response to prepare a Set of values for PARENT_CONNECTION_PROCESS_QUALIFIED_NAME for each returned asset in response, This Set can be iterated to call deleteConnectionProcess